File: src/main/cpp/subsystems/subsystem_Intake.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "subsystems/subsystem_Intake.h"
#include <frc/smartdashboard/SmartDashboard.h>

subsystem_Intake::subsystem_Intake(): 
                                      m_IntakeMotor{IntakeConstants::IntakeMotorID, rev::CANSparkMax::MotorType::kBrushless} 
{
  m_IntakeMotor.SetSmartCurrentLimit(IntakeConstants::IntakeMotorCurrentLimit);
}

bool subsystem_Intake::CheckGamepieceIntake(){
  frc::SmartDashboard::SmartDashboard::PutNumber("Output Current", m_IntakeMotor.GetOutputCurrent());
   return (m_intakeState == IntakeConstants::cubeIntaking || m_intakeState == IntakeConstants::coneIntaking)
  && m_IntakeMotor.GetOutputCurrent() > IntakeConstants::normalOutputCurrent;
}

IntakeConstants::intakeState subsystem_Intake::getState(){
  return m_intakeState;
}

void subsystem_Intake::setState(IntakeConstants::intakeState state){
  m_intakeState = state;
}



// This method will be called once per scheduler run
void subsystem_Intake::Periodic() {
  switch(m_intakeState){
    case (IntakeConstants::coneIntaking):
    frc::SmartDashboard::SmartDashboard::PutString("Intake State", "coneIntaking");
    m_IntakeMotor.Set(0.5);
      break;
    case (IntakeConstants::cubeOuttaking):
    frc::SmartDashboard::SmartDashboard::PutString("Intake State", "cubeOuttaking");
      m_IntakeMotor.Set(0.5);
      break;
    case (IntakeConstants::cubeIntaking):
    frc::SmartDashboard::SmartDashboard::PutString("Intake State", "cubeIntaking");
    m_IntakeMotor.Set(-0.5);
      break;
    case (IntakeConstants::coneOuttaking):
    frc::SmartDashboard::SmartDashboard::PutString("Intake State", "coneOuttaking");
      m_IntakeMotor.Set(-0.5);
      break;
    case (IntakeConstants::hasCone):
    frc::SmartDashboard::SmartDashboard::PutString("Intake State", "hasCone");
    m_IntakeMotor.Set(0);
    break;
    case (IntakeConstants::hasCube):
    frc::SmartDashboard::SmartDashboard::PutString("Intake State", "hasCube");
    m_IntakeMotor.Set(0);
    break;
    case (IntakeConstants::noGamepiece):
    frc::SmartDashboard::SmartDashboard::PutString("Intake State", "noGamepiece");
    m_IntakeMotor.Set(0);
    break;
  }
}

File: src/main/java/frc/robot/subsystems/subsystem_Intake.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import edu.wpi.first.wpilibj2.command.CommandBase;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants.IntakeConstants;
import frc.robot.Constants.IntakeConstants.intakeState;

import com.revrobotics.CANSparkMax;
import com.revrobotics.CANSparkMax.IdleMode;
import com.revrobotics.CANSparkMaxLowLevel.MotorType;

public class subsystem_Intake extends SubsystemBase {
  /** Creates a new subsystem_Intake. */
  private intakeState m_IntakeState = intakeState.noGamepiece;
  private CANSparkMax m_IntakeMotor;

  public subsystem_Intake() {
    m_IntakeMotor = new CANSparkMax(IntakeConstants.intakeMotorID, MotorType.kBrushless);
    m_IntakeMotor.setSmartCurrentLimit(IntakeConstants.intakeMotorCurrentLimit);
  }

  public boolean checkGamepieceIntake(){
    return (m_IntakeState == intakeState.cubeIntaking || m_IntakeState == intakeState.coneIntaking)
    && m_IntakeMotor.getOutputCurrent() > IntakeConstants.normalOutputCurrent;
  }

  public intakeState getState(){
    return m_IntakeState;
  }

  public void setState(intakeState state){
    m_IntakeState = state;
    switch (m_IntakeState){
      case coneIntaking:
      case cubeOuttaking:
        m_IntakeMotor.set(-0.5);
        break;
      case coneOuttaking:
      case cubeIntaking:
        m_IntakeMotor.set(0.5);
        break;
      case hasCone:
      case hasCube:
      case noGamepiece:
        m_IntakeMotor.set(0);
        break;
      default:
        m_IntakeMotor.set(0);
    }
  }
  
  public void setBrakeMode(){
    m_IntakeMotor.setIdleMode(IdleMode.kBrake);
  }

  public void setCoastMode(){
    m_IntakeMotor.setIdleMode(IdleMode.kCoast);
  }
  
  @Override
  public void periodic() {
    // This method will be called once per scheduler run
  }

  public CommandBase toggleIntake(){
    // implicitly require `this`

    return m_IntakeMotor.get()> 0.1 ? this.runOnce(() -> m_IntakeMotor.set(0)) : 
    this.runOnce(() -> m_IntakeMotor.set(0.5));
}

}
